### PR TITLE
Handle IForToLoopOperation over object in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb [hunk 1]
```diff
@@ -6448,5 +6448,1166 @@ Block[B27] - Exit
...
+                    IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Object, IsImplicit) (Syntax: 'For i = ini ... Next')
+                      Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                        (WideningValue)
+                      Operand: 
+                        ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1, IsImplicit) (Syntax: 'For i = ini ... Next')
+                    InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                    OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                  IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: LoopForResult) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: 'i')
+                    ILocalReferenceOperation:  (IsDeclaration: True) (OperationKind.LocalReference, Type: System.Object, IsImplicit) (Syntax: 'i')
```
Comment (mavasani) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: I presume this is the same anonymous local that is declared in the enclosing region `R1`? Wondering why the name for this is blank, while the dump above contains `<anonymous local>`

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -2981,33 +2988,40 @@ ImmutableArray<IArgumentOperation> makeArguments(Lazy<ImmutableArray<IArgumentOp
...
-            var @break = GetLabeledOrNewBlock(operation.ExitLabel);
+            BasicBlock @continue = GetLabeledOrNewBlock(operation.ContinueLabel);
+            BasicBlock @break = GetLabeledOrNewBlock(operation.ExitLabel);
+            BasicBlock checkConditionBlock = new BasicBlock(BasicBlockKind.Block);
+            BasicBlock bodyBlock = new BasicBlock(BasicBlockKind.Block);
 
             if (haveLocals)
             {
-                EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.Locals, locals: operation.Locals));
+                EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.Locals, locals: locals));
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Nevermind, I see that it's out of order.

---
In reply to: [188116692](https://github.com/dotnet/roslyn/pull/26803#discussion_r188116692) [](ancestors = 188116692)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -3033,8 +3047,88 @@ void initializeLoop()
...
-                    // PROTOTYPE(dataflow): Handle loops over object
-                    throw new NotImplementedException();
+                    // For i as Object = 3 To 6 step 2
+                    //    body
+                    // Next
+                    //
+                    // becomes ==>
+                    //
+                    // {
+                    //   Dim loopObj        ' mysterious object that holds the loop state
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >mysterious [](start = 46, length = 10)

Probably not the right word :)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Probably not the right word :)

Works for VB lowering, where I copied this comment from.

---
In reply to: [188118434](https://github.com/dotnet/roslyn/pull/26803#discussion_r188118434) [](ancestors = 188118434)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: > I presume this is the same anonymous local that is declared in the enclosing region R1? Wondering why the name for this is blank, while the dump above contains <anonymous local>

This is a difference in how the tree serialization and graph serialization are implemented
OperationTreeVerifier.VisitLocalReference:
```
LogString($": {operation.Local.Name}");
```

ControlFlowGraphVerifier.printLocals:
```
stringBuilder.Append($" [{local.ToTestDisplayString()}]");
```

A change to OperationTreeVerifier.VisitLocalReference is going to "break" a lot of tests.

---
In reply to: [188112480](https://github.com/dotnet/roslyn/pull/26803#discussion_r188112480) [](ancestors = 188112480)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -3033,8 +3047,88 @@ void initializeLoop()
...
+                    //       body
+                    //
+                    //   continue:
+                    //       ' helper updates loop state and tells if we need to do another iteration.
+                    //       if ObjectFlowControl.ForLoopControl.ForNextCheckObj(ctrl, loopObj, ref ctrl) 
+                    //                               GoTo start
+                    // }
+                    // exit:
+
+                    IOperation condition;
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 2) - Comments can be addressed in a follow up.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

I'll se what I can do (in a follow-up PR)

---
In reply to: [188128275](https://github.com/dotnet/roslyn/pull/26803#discussion_r188128275) [](ancestors = 188128275)

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need a second review.

**Review by cston — APPROVED** [review]

**Pull request merged**